misc/helper-scripts/rosmsg_to_asn1.py

```diff
@@ -0,0 +1,417 @@
+    try:
+        msg_dict = dict(
+            type=obj._type, slots=obj.__slots__, slot_types=obj._slot_types,
+            full_text=obj._full_text)
+    except KeyError as e:
+        logging.error(
+            "Could not read the information of message {}.".format(msg))
+        return None
+    logging.debug("Dictionary for message {0} is:".format(msg))
+    logging.debug(pprint.pformat(msg_dict))
+    return msg_dict
+
+
+def find_field_type(slot_type):
+    """
+    Finds the field type of a field. If it is an array it returns the data type
+    of the array along the array type and its length.
+    :param slot_type: The type of the slot whose field type is to be found.
+    :return slot_type: The base data type of the slot (if it is an array).
+    :return field_type: The field_type of the slot.
+    :return length: The length of the array.
+    """
+    bracket = "["
+    length = 1
+    field_type = "simple"
+    idx = slot_type.find(bracket)
+    if idx == -1:
+        return slot_type, field_type, length
+    else:
+        if slot_type[idx + 1] == "]":
+            field_type = "variable-length"
+        else:
+            field_type = "fixed-length"
+            idx_end = slot_type.find("]")
+            str_length = slot_type[idx+1:idx_end]
+            length = int(str_length)
+        slot_type = slot_type[0:idx]
+    return slot_type, field_type, length
+
+
+def find_ASN1_type(slot_type):
+    """
+    Given the ROS type of a field finds and returns the equivalent ASN.1 type.
+    If the type is not primitive a flag is set so that a dependency is
+    generated.
+    :param slot_type: The ROS field type.
+    :return asn1_type: The ASN.1 equivalent type of the ROS type.
+    :return lib_asn: The library the ASN.1 type belongs to.
+    :return primitive: Flag to denote if the field type is primitive.
+    """
+    primitive = False
+    if slot_type in ros_primitive_type:
+        asn1_type = primitive_types_map[slot_type]
+        lib_asn = libraries[asn1_type]
+        primitive = True
+    else:
+        composed_type = slot_type.split("/")
+        if len(composed_type) == 1:
+            logging.warn("The length of the composed_type is 1!")
+        asn1_type = composed_type[1]
+        # primitive_types_map[slot_type] = asn1_type
+        lib_asn = composed_type[1]+'-Types'
+        # libraries[asn1_type] = lib_asn
+        primitive = False
+    return asn1_type, lib_asn, primitive
+
+
+def generate_libraries_string(import_libraries):
+    """
+    Generate a string from a set of libraries to be imported to the ASN.1 type.
+    :param import_libraries: The libraries to be imported.
+    :return s: The libraries names in string format.
+    """
+    s = "IMPORTS "
+    comma = ""
+    for key in import_libraries:
+        num_types = len(import_libraries[key])
+        for t in import_libraries[key]:
+            s += comma + t
+            comma = ", "
+        s += " FROM " + key + " "
+        comma = ""
```

Comment:
If I read this correctly, then for `{'key1' : ['a', 'd'], 'key2':['b', c'] }` this would generate:
```
IMPORTS a, d FROM key1 b, c FROM key2
```
is this expected behaviour?

Would it make sense to generate one IMPORTS per key? Is that allowed by ASN1?

Some magic could make this 'simpler':

```
s = ""
for key in import_libraries:
  import_list = "{}".format(list(import_libraries[key])) # includes brackets [1:-1] removes the brackets
  s += "IMPORTS {} FROM {};\n".format(import_list[1:-1]), key))
return s
```
or
```
s = "IMPORTS"
for key in import_libraries:
  import_list = "{}".format(list(import_libraries[key])) # includes brackets [1:-1] removes the brackets
  s += " {} FROM {}".format(import_list[1:-1]), key))
s += ";\n"
return s
```

Reply:
AFAIK ASN.1 wants only one IMPORTS statement. Used your second method.